### PR TITLE
Add location dux module

### PR DESCRIPTION
**Pull request by prodhan29**
- add ducks module location-based-breadcrumb 
- test code for duck module

### CODE REVIEW

File: clients/core/src/store/ducks/locations.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
+import { AnyAction, Store } from 'redux';
+import SeamlessImmutable from 'seamless-immutable';
+
+/** The reducer name */
+export const reducerName = 'locations';
+
+/** A single node for location hierarchy */
+export interface LocationNode {
```
[the remark this answers is not in the document]
Comment (prodhan29) [reply clients/core/src/store/ducks/locations.ts]: yes. here is an example. [https://github.com/prodhan29/breadcrum-location/blob/real-location-data/src/sample-data.js](url)

File: clients/core/src/store/ducks/tests/location.test.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        store.dispatch(setLocations(fixtures.location));
+        expect(getLocationHierarchy(store.getState())).toEqual(fixtures.location);
+    });
+
+    it(' spy on location hierarchy', () => {
+        const spyOnLocationHierarchy = jest.spyOn(locationDucksModule, 'setLocations');
+        store.dispatch(setLocations(fixtures.location));
+        expect(spyOnLocationHierarchy).toHaveBeenCalledTimes(1);
+        expect(spyOnLocationHierarchy).toHaveBeenCalledWith(fixtures.location);
+    });
```
Comment (p-netm) [inline clients/core/src/store/ducks/tests/location.test.ts]: What exactly is being tested here?

File: clients/core/src/store/ducks/tests/fixtures.ts [hunk 3]
```diff
@@ -335,5 +336,39 @@ export const household3: Household = {
     _rev: 'v1',
 };
 
+export const location: LocationHierarchy = {
```
Comment (p-netm) [inline clients/core/src/store/ducks/tests/fixtures.ts]: can we include a few more fixtures for this, or a more comprehensive one

File: clients/core/src/store/ducks/locations.ts [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+export const reducerName = 'locations';
+
+/** Inteface for Location */
+export interface Location {
+    id: string;
+    label: string;
+    node: {
+        locationId: string;
+        name: string;
+        parentLocation?: {
```
Comment (p-netm) [inline clients/core/src/store/ducks/locations.ts]: Breaking down this interface could make it easier to use recursion in declaring this interface. for instance: 
```typescript
interface Parent {
  name: string;
  child?: Parent
}

const matryoshka: Parent = {
  name: 'grandparents',
  child: {
    name: 'parents',
    child: {
      name: 'child',
    }
  }
}
```

File: clients/core/src/store/ducks/tests/fixtures.ts (hunk 3 above)
Comment (prodhan29) [reply clients/core/src/store/ducks/tests/fixtures.ts]: make it more comprehensive

File: clients/core/src/store/ducks/locations.ts (hunk 4 above)
Comment (prodhan29) [reply clients/core/src/store/ducks/locations.ts]: Done.

File: clients/core/src/store/ducks/tests/location.test.ts (hunk 2 above)
Comment (prodhan29) [reply clients/core/src/store/ducks/tests/location.test.ts]: making sure that action is called not more than once.

File: clients/core/src/store/ducks/tests/fixtures.ts (hunk 3 above)
Comment (prodhan29) [reply clients/core/src/store/ducks/tests/fixtures.ts]: update the location fixture.

File: clients/core/src/store/ducks/tests/location.test.ts (hunk 2 above)
Comment (p-netm) [reply clients/core/src/store/ducks/tests/location.test.ts]: my problem with this test is that, `setLocations` could be any function, doing God knows what, but still this test would pass.

Comment (p-netm) [reply clients/core/src/store/ducks/tests/location.test.ts]: it might seem unlikely to occur in production, but maybe also add a test that sets locations on a non-empty state.

File: clients/core/src/store/ducks/tests/location.test.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+        expect(getLocationHierarchy(store.getState())).toEqual({});
+    });
+
+    it(' set locations correctly', () => {
+        store.dispatch(setLocations(fixtures.location));
+        expect(getLocationHierarchy(store.getState())).toEqual(fixtures.location);
+    });
+
+    it(' can set locations on a non empty state', () => {
+        store.dispatch(setLocations(null));
```
Comment (p-netm) [inline clients/core/src/store/ducks/tests/location.test.ts]: instead of using null add one more fixture for this.

Comment (prodhan29) [reply clients/core/src/store/ducks/tests/location.test.ts]: okay

Comment (prodhan29) [reply clients/core/src/store/ducks/tests/location.test.ts]: @p-netm done
